### ISSUE TITLE
current_thread() becomes "dummy" thread during shutdown

### ISSUE THREAD

**Issue by pitrou** [python/cpython#75697]
BPO | [31516](https://bugs.python.org/issue31516)
--- | :---
Nosy | @tim-one, @pitrou, @vstinner, @asvetlov, @serhiy-storchaka, @miss-islington
PRs | <li>python/cpython#3673</li><li>python/cpython#3856</li><li>python/cpython#8052</li><li>python/cpython#9813</li><li>python/cpython#9814</li>
Files | <li>[mainthread.py](https://bugs.python.org/file47152/mainthread.py "Uploaded as text/plain at 2017-09-19.11:33:54 by @pitrou")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-10-02.15:20:39.908>
created_at = <Date 2017-09-19.11:33:54.534>
labels = ['3.7', 'type-bug', 'library']
title = 'current_thread() becomes "dummy" thread during shutdown'
updated_at = <Date 2018-10-12.08:54:24.047>
user = 'https://github.com/pitrou'
```

bugs.python.org fields:
```python
activity = <Date 2018-10-12.08:54:24.047>
actor = 'miss-islington'
assignee = 'none'
closed = True
closed_date = <Date 2017-10-02.15:20:39.908>
closer = 'pitrou'
components = ['Library (Lib)']
creation = <Date 2017-09-19.11:33:54.534>
creator = 'pitrou'
dependencies = []
files = ['47152']
hgrepos = []
issue_num = 31516
keywords = ['patch']
message_count = 10.0
messages = ['302515', '302517', '302637', '302641', '303537', '303542', '303543', '327572', '327573', '327575']
nosy_count = 6.0
nosy_names = ['tim.peters', 'pitrou', 'vstinner', 'asvetlov', 'serhiy.storchaka', 'miss-islington']
pr_nums = ['3673', '3856', '8052', '9813', '9814']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue31516'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by pitrou** [comment]
Attaching reproducer.  During interpreter shutdown, threading.main_thread() gets marked "stopped" (while it's obviously still running), while threading.current_thread() returns a DummyThread instance with the same ident as the main thread.

```sh
$ ./python mainthread.py 
-- before shutdown --
GC in thread <_MainThread(MainThread, started 140359122872064)>
main thread is <_MainThread(MainThread, started 140359122872064)>
-- during shutdown --
GC in thread <_DummyThread(Dummy-1, started daemon 140359122872064)>
main thread is <_MainThread(MainThread, stopped 140359122872064)>
```

**Comment by pitrou** [comment]
The main thread being marked "stopped" at shutdown might be an obscure feature (as hinted in the comments for threading.\_shutdown()).  However, current_thread() should definitely not return a separate DummyThread instance.

**Comment by pitrou** [comment]
Tim, any opinion on this one?  Do you agree it's an actual bug (if a bit obscure)?

**Comment by tim-one** [comment]
Ya, it's clearly best for `current_thread()` to deliver consistent results.

**Comment by pitrou** [comment]
New changeset 1023dbbcb7f05e76053486ae7ef7f73b4cdc5398 by Antoine Pitrou in branch 'master':
bpo-31516: current_thread() should not return a dummy thread at shutdown (bpo-3673)
https://github.com/python/cpython/commit/1023dbbcb7f05e76053486ae7ef7f73b4cdc5398

**Comment by pitrou** [comment]
New changeset ac6245a31f9a757db0520722c592cb7fdcb55eb0 by Antoine Pitrou (Miss Islington (bot)) in branch '3.6':
[3.6] bpo-31516: current_thread() should not return a dummy thread at shutdown (GH-3673) (bpo-3856)
https://github.com/python/cpython/commit/ac6245a31f9a757db0520722c592cb7fdcb55eb0

**Comment by pitrou** [comment]
I have now pushed the fix to master and 3.6.  Hopefully this won't break anyone's code...
